api/src/main/java/org/projectnessie/tools/catalog/migration/api/CatalogMigrator.java

```diff
@@ -0,0 +1,277 @@
+                  + "Use the tables from target catalog."));
+    }
+
+    printWriter.println(String.format("%nStarted %s ...", operation));
+    List<TableIdentifier> registeredTableIdentifiers = new ArrayList<>();
+    List<TableIdentifier> failedToRegisterTableIdentifiers = new ArrayList<>();
+    List<TableIdentifier> failedToDeleteTableIdentifiers = new ArrayList<>();
+    AtomicInteger counter = new AtomicInteger();
+    identifiers.forEach(
+        tableIdentifier -> {
+          boolean isRegistered =
+              registerTable(
+                  catalogMigratorParams.sourceCatalog(),
+                  catalogMigratorParams.targetCatalog(),
+                  registeredTableIdentifiers,
+                  failedToRegisterTableIdentifiers,
+                  tableIdentifier);
+
+          // HadoopCatalog dropTable will delete the table files completely even when purge is
+          // false. So, skip dropTable for HadoopCatalog.
+          boolean deleteTableFromSourceCatalog =
+              !(catalogMigratorParams.sourceCatalog() instanceof HadoopCatalog)
+                  && isRegistered
+                  && deleteEntriesFromSourceCatalog;
+          try {
+            if (deleteTableFromSourceCatalog) {
+              boolean isDropped =
+                  catalogMigratorParams.sourceCatalog().dropTable(tableIdentifier, false);
+              if (!isDropped) {
+                failedToDeleteTableIdentifiers.add(tableIdentifier);
+              }
+            }
+          } catch (Exception exception) {
+            failedToDeleteTableIdentifiers.add(tableIdentifier);
+            LOG.warn("Failed to delete the table after migration {}", tableIdentifier, exception);
+          }
+
+          int count = counter.incrementAndGet();
+          if (count % 100 == 0) {
+            printWriter.println(
+                String.format(
+                    "%nAttempted %s for %d tables out of %d tables.",
+                    operation, count, identifiers.size()));
+          }
+        });
+    printWriter.println(String.format("%nFinished %s ...", operation));
+
+    CatalogMigrationResult result =
+        ImmutableCatalogMigrationResult.builder()
+            .registeredTableIdentifiers(registeredTableIdentifiers)
+            .failedToRegisterTableIdentifiers(failedToRegisterTableIdentifiers)
+            .failedToDeleteTableIdentifiers(failedToDeleteTableIdentifiers)
+            .build();
+
+    if (!result.failedToRegisterTableIdentifiers().isEmpty()) {
+      writeToFile(
+          pathWithOutputDir(catalogMigratorParams.outputDirPath(), FAILED_IDENTIFIERS_FILE),
+          result.failedToRegisterTableIdentifiers());
+    }
+    if (!result.failedToDeleteTableIdentifiers().isEmpty()) {
+      writeToFile(
+          pathWithOutputDir(catalogMigratorParams.outputDirPath(), FAILED_TO_DELETE_AT_SOURCE_FILE),
+          result.failedToDeleteTableIdentifiers());
+    }
+
+    result.printSummary(
```

Comment:
You've only moved code from one place to another. But the "API" should not print directly.

Reply:
Users can wrap the `printWriter` with `OutputStream.nullOutputStream()` to skip printing the outputs.  

As the same API is used for CLI calls, we need to provide intermediate output on the console. Splitting the register table and intermediate output writing can bring additional overhead to CLI logic and can be complex.

Reply:
checking about splitting the identifiers into batches and removing `printWriter` from api module.